### PR TITLE
Fix timestamp semantics

### LINKED ISSUE #37: TIMESTAMP behaviour does not match sql standard
## Problem description

See below.

## Roadmap

- [x] use reasonable temporal representation in tests (testability)
- [x] feature config and session toggles
- [x] use non-UTC zone as default in test and better default JVM zone (testability) (= https://github.com/prestodb/presto/issues/10078, more details there)
- [x] (https://github.com/prestodb/presto/pull/10193) engine semantics fixes 
- [x] all casts from TIME\* to TIMESTAMP\* -- they fill date fields with `current_date`, not 1970-01-01 (see https://github.com/prestodb/presto/issues/10103#issuecomment-408256717 & following)
- [x] Remove support for political time zones from `TIME WITH TIME ZONE` (= https://github.com/prestosql/presto/issues/191, more details there)
- [x] Add product-test environments with Hive 3.1 (or better) (https://github.com/prestosql/presto/issues/190)
- [ ] Hive changes: Hive `TIMESTAMP` should be matched to Presto `TIMESTAMP WITH TIME ZONE` (??)
   * [x] ORC improvement: https://github.com/prestosql/presto/pull/212
   * however https://issues.apache.org/jira/browse/HIVE-12191 & https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types#LanguageManualTypes-TimestampstimestampTimestamps look as if Hive has changed / is going to change it's timestamp semantics, so the conversion might be hive's version dependent? References: https://issues.apache.org/jira/browse/HIVE-16614?focusedCommentId=16425288&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16425288, https://issues.apache.org/jira/browse/HIVE-12192
   * must-read: https://cwiki.apache.org/confluence/display/Hive/Different+TIMESTAMP+types
- [x] Hive changes in file handlers: ORC & Parquet (especially for predicate pushdown, e.g. https://github.com/prestodb/presto/pull/10181#issuecomment-373869651), others
- [ ] Raptor changes (...?)
- [ ] JDBC connectors changes: 
   * [x] reading/writing -- general changes (https://github.com/prestosql/presto/pull/495)
   * [x] tests for Postgres (https://github.com/prestosql/presto/pull/495)
   * [ ] tests for MySQL
   * [ ] tests for Redshift (https://github.com/prestosql/presto/issues/497)
   * [ ] tests for SQLServer
   * [x] tests for predicate pushdown (https://github.com/prestosql/presto/issues/496)
- all other connectors:
  - [x] Cassandra (https://github.com/prestosql/presto/issues/2269)
  - [x] Elasticsearch (https://github.com/prestosql/presto/issues/3742)
  - [ ]  ................ (to be enumerated)
- [x] Presto JDBC changes (https://github.com/prestosql/presto/pull/307, etc.)
- [ ] Q:  what is the complete set of implicit coercions we are going to have between TIME, TIME WITH TZ, TIMESTAMP, TIMESTAMP WITH TZ?
- [x] Q: should `from_unixtime`, `to_unixtime` continue to work with `TIMESTAMP`?
- [x] enable "new" correct semantics by default
- [ ] get rid of `catch Exception` for checking if input is parsable in `DateTimeUtils`

(Note: since GH doesn't send updates for edits, when adding a bullet please be sure sure to add a comment too.)

## Original problem description

It seems like meaning of `TIMESTAMP` and `TIMESTAMP WITH TIMEZONE` datatypes in Presto is totally not what is specified by SQL standard (and what other databases do).

This is my understanding of SQL 2003 standard (4.6.2 Datetimes):

`TIMESTAMP WITH TIMEZONE` represents absolute point in time. Typically databases store it internally as seconds since epoch in some fixed timezone (usually UTC). When querying `TIMESTAMP WITH TIMEZONE` data the values are presented to user in session timezone (yet session timezone is used just for presentation purposes).

`TIMESTAMP` does not represent specific point in time, but rather a reading of a wall clock+calendar. Selecting values from `TIMESTAMP` column should return same result set no matter what is the client's session timezone.

While Presto semantics is different:

`TIMESTAMP` seems to do what `TIMESTAMP WITH TIMEZONE` should.

`TIMESTAMP WITH TIMEZONE` encodes explicit timezone information to each value stored in table. The sql standard does not define a type like that. But it does not seem very practical. Assuming that values selected from `TIMESTAMP WITH TIMEZONE` are presented to user in session timezone anyway, the per-row timezone information can be stripped away and all values can be stored in some arbitrary fixed timezone (e.g. UTC).

Please comment on the semantics. It seems wrong. Why the choice - as it is hard to believe that it was not done intentionally.

### 1
@losipiuk I agree with you that Timestamp w/o TZ in Presto is broken. I do NOT agree that Timestamp w TZ should behave like Instant. I believe it should also have an associated time zone. (In other word, I believe Timestamp w TZ is implemented correctly today.) Below is an excerpt of something I wrote early last year that summarizes the current behavior and my understanding.

-------

To summarize how things work today:

* Timestamp w TZ = DateTime in joda = ZonedDateTime in java8
* Timestamp w/o TZ = Instant in joda = Instant in java8. 
  * In other words, Timestamp w/o TZ represents an instant in time, they just don't have a timezone associated with them.
  * When you print them out for human consumption, you need to resort to user's session time zone.
  * When you turn it into a Timestamp w TZ, you just stick that time zone to it without changing the instant.

The way I understand it

* Timestamp w TZ = DateTime in joda = ZonedDateTime in java8
* Timestamp w/o TZ = LocalDateTime in joda = LocalDateTime in java8.
  * In other words, Timestamp w/o TZ represents a specific Year/Month/Day/Hour/Minute/Second. But it doesn't represent a specific instant. (Side note: this concept does not need expensive representation. It can still be represented as millis/nanos since epoch, observing only chronology rules and no tz rules)
  * When you print it out, it should be printed out as is.
  * If you want to turn it into a Timestamp w TZ, you need to resort to the user's session time zone for the missing piece of information (and this is not always possible because Timestamp w TZ has gaps).

Here is the reason I believe the first understanding is inconsistent. I can only think of one possible interpretation for the other 3 concepts:
* Time w TZ = no single class correspondence in joda = OffsetTime in java8
  * Offset and Zone is different in that only constant offset exists for OffsetX. Political zone is not allowed.
  * ZonedTime doesn't make sense and doesn't exist. ZonedDateTime and OffsetDateTime both exists.
* Time w/o TZ = LocalTime in joda = LocalTime in java8
* Date = LocalDate in joda = LocalDate in java8

Note here the inconsistency between interpretation of Timestamp w/o TZ and Time w/o TZ if we adopt the first interpretation of Timestamp w/o TZ (Instant vs LocalTime). Whereas under the second interpretation, it will be consistent (LocalDateTime vs LocalTime).

I went to SQL spec for the definitive answer:
* Abbreviations
  * SV is the source value, TV is the target value
  * UTC is the UTC component of SV or TV (if and only if the source or target has time zone)
  * TZ is the timezone displacement of SV or TV (if and only if the source or target has time zone)
  * STZD is the SQL-session default time zone displacement
* To convert Timestamp w/o TZ to Timestamp w/ TZ: TV.UTC = SV - STZD; TV.TZ = STZD
* To convert Timestamp w/ TZ to Timestamp w/o TZ: TV = SV.UTC + SV.TZ

I believe these two rules proves that SQL spec agrees with my interpretation. Let's consider cast from Timestamp w/o TZ to Timestamp w/ TZ
* Cast from Timestamp w/o TZ
  * Given, SV = 3600000 millis (1 hour)
* Cast to Timestamp w TZ in MPK
  * STZD = America/Los_Angeles
  * TV.UTC = 3600000 - (-28800000) = 32400000, TV.TZ=America/Los_Angeles
  * When written out in human format, TV is: 1970-01-01 01:00:00 America/Los_Angeles, TV.UTC is 1970-01-01 09:00:00
* Cast to Timestamp w TZ in Shanghai
  * STZD = Asia/Shanghai
  * TV.UTC = 3600000 - 28800000 = -25200000, TV.TZ=Asia/Shanghai
  * When written out in human format, TV is: 1970-01-01 01:00:00 Asia/Shanghai, TV.UTC is 1969-12-31 17:00:00

Under first interpretation, these two cast should have yield results that are equal. Under second interpretation, they would produce different result. The rule in SQL spec produces two different results.

Lastly, a side note from me. Both interpretation can produce results that is dependent on user session time zone:

* Under current Presto interpretation of Timestamp w/o TZ:
  * Same storage representation. Let's say 60000 milliseconds.
  * Prints out as 1969-12-31 16:01:00 in MPK, and 1970-01-01 08:01:00 in China. They are different.
  * When cast to Timestamp w TZ, become 1969-12-31 16:01:00 America/Los_Angeles in MPK, and 1970-01-01 08:01:00 Asia/Shanghai in China. They are equal.
* Under SQL spec interpretation of Timestamp w/o TZ:
  * Same storage representation. Let's say 60000 milliseconds.
  * Prints out as 1970-01-01 00:01:00 in MPK, and 1970-01-01 00:01:00 in China. They are the same.
  * When cast to Timestamp w TZ, become 1970-01-01 00:01:00 America/Los_Angeles in MPK, and 1970-01-01 00:01:00 Asia/Shanghai in China. They are different.

Under the SQL spec, cast from timestamp w/o TZ to timestamp w/ TZ can produce different results based on user time zone. As a result, I guess this cast probably should **NOT** have been implicit.

### 2
@losipiuk, @dain, and I reached agreement:
* `Timestamp with Timezone` in Presto is implemented properly today (like `DateTime` in joda, `ZonedDateTime` in jdk8).
* `Timestamp` in Presto is like `Instant` in joda/jdk8 today. It should be like `LocalDateTime` in joda/jdk8.
* Extracting hour from `2016-01-01 12:00:00 <TZ>` should return 12 no matter what `<TZ>` is put in template.
* As part of fixing `Timestamp` in Presto, we should remove implicit coercion from `Timestamp` to `Timestamp with Timezone` because the result value is environment dependent.

## Notes

Ticket migrated from https://github.com/prestodb/presto/issues/7122, https://github.com/prestodb/presto/issues/10326

### LINKED ISSUE #3742: Fix timestamp mapping in Elasticsearch for non-legacy timestamp
https://github.com/prestosql/presto/blob/13a3c523f2739a2373c251697f6826daaf193266/presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/TimestampDecoder.java#L82

should use UTC for non-legacy timestamp semantics.

Relates to https://github.com/prestosql/presto/issues/3741

### PR DESCRIPTION
**Pull request by dain**
Fixes #37 
Fixes #3742

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java [hunk 1]
```diff
@@ -159,7 +114,7 @@ private static long castToShortTimestamp(int precision, String value, Function<S
         return epochSecond * MICROSECONDS_PER_SECOND + rescale(fractionValue, actualPrecision, 6);
     }
 
-    private static LongTimestamp castToLongTimestamp(int precision, String value, Function<String, ZoneId> zoneId)
+    public static LongTimestamp castToLongTimestamp(int precision, String value)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java:117]: Same here

File: presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java [hunk 2]
```diff
@@ -71,47 +57,17 @@ public static long castToShort(@LiteralParameter("p") long precision, ConnectorS
...
-        return castToLongTimestamp(precision, value, timezone -> {
-            if (timezone == null) {
-                return timeZoneKey.getZoneId();
-            }
-            return ZoneId.of(timezone);
-        });
-    }
-
-    private static long castToShortTimestamp(int precision, String value, Function<String, ZoneId> zoneId)
+    public static long castToShortTimestamp(int precision, String value)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java:70]: Why public? This is supposed to be an implementation detail of this class.

**Comment by electrum** [comment]
I previously reviewed this offline with @dain as the change was developed.

File: presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java (hunk 2 above)
Comment (dain) [reply presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java:70]: They replace the old legacy method used in tests.  I will mark both as visible for testing.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/AbstractTestHiveLocal.java [hunk 3]
```diff
@@ -70,7 +70,8 @@ public void initialize()
                         .build());
 
         HiveConfig hiveConfig = new HiveConfig()
-                .setTimeZone("America/Los_Angeles");
+                .setParquetTimeZone("America/Los_Angeles")
+                .setRcfileTimeZone("America/Los_Angeles");
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/AbstractTestHiveLocal.java:74]: Why this zone in particular?

**Pull request merged**